include/io/mdarray.hpp

```diff
@@ -0,0 +1,274 @@
+     * mdarray<int> x(2); // rank 2
+     * x.at({0, 0});      // get the value at (0, 0) in a 2D array.
+     *
+     * @param n Index list
+     * @return reference 
+     */
+    const_reference at(const ilist& n) const
+    {
+        return this->m_data.at(this->index(n));
+    }
+
+    void insert(const ilist& n, value_type value)
+    {
+        for (size_type i = 0; i < this->m_shape.size(); i++) {
+            if (n[i] >= this->m_shape[i])
+                throw std::out_of_range("out of range");
+        }
+
+        size_type index = this->index(n);
+    
+        if (index > this->max_size())
+            throw std::out_of_range("out of range");
+        
+        if (index > this->size() || this->size() == 0)
+            this->m_data.resize(index + 1);
+    
+        this->m_data.at(index) = value;
+    }
+
+    void insert(std::initializer_list<std::pair<std::initializer_list<size_type>, value_type>> args)
+    {
+        for (const auto& arg : args) {
+            this->insert(arg.first, arg.second);
+        }
+    }
+
+    /**
+     * Allocate this mdarray based on the indices given.
+     * 
+     * @example
+     * mdarray<double> x(3);   // rank 3
+     * x.allocate({5, 10, 10}) // Allocates 5 * 10 * 10 elements
+     * @param n 
+     */
+    void allocate(const ilist& n)
+    {
+        this->m_data.resize(this->index(n));
+    }
+
+    /**
+     * Get the (total) size of this mdarray
+     * (aka the total number of elements).
+     * 
+     * @return size_type
+     */
+    size_type size() const noexcept
+    {
+        return this->m_data.size();
+    }
+
+    size_type max_size() const noexcept
+    {
+        size_type max = 1;
+        for (const auto& v : this->m_shape) {
+            max *= v;
+        }
+        return max;
+    }
+
+    iterator begin() const noexcept {
+        return iterator(*this, 0);
+    }
+
+    iterator end() const noexcept {
+        return iterator(*this, this->m_data.size());
+    }
+
+    /**
+     * Get the rank of this mdarray.
+     * 
+     * @return size_type 
+     */
+    size_type rank() const noexcept
+    {
+        return this->m_shape.size();
+    }
+
+    /**
+     * Get the shape of this mdarray.
+     * 
+     * @return ilist
+     */
+    ilist shape() const noexcept
+    {
+        return this->m_shape;
+    }
+
+    /**
+     * Index a multi-dimensional set of indices to a single address index.
+     * 
+     * @param n 
+     * @return size_type 
+     */
+    size_type index(const ilist& n) const
+    {
+        size_type index = 0;
+        size_type stride = 1;
+        for (size_type k = 0; k < this->rank(); k++) {
+            index += n[k] * stride;
+            stride *= this->m_shape[k];
+        }
+    
+        return index;
+    }
+
+    /**
+     * Retrieve the multi-dimensional index from a given address index.
+     * 
+     * @param idx Address index (aka flat index)
+     * @return ilist 
+     */
+    ilist deindex(size_type idx) const
```

Comment:
In this case, we should definitely be taking an 'output argument' that we'll write the result into, rather than constructing and returning a temporary `vector`. The use in `iterator::mdindex()` called in a loop by `mdarray_putvar` will be painful otherwise, since it will do that for every single element